Documentation/AdditionalFunctionality/PushNotifications.md

```diff
@@ -0,0 +1,164 @@
+
+    @Environment(\.notificationCenter) var notificationCenter: NotificationCener
+
+    func handle() -> EventLoopFuture<Void> {
+        notificationCenter.register(device, to: user);
+    }
+}
+```
+
+Furthermore, the `NotificationCenter` allows the removal and editing of `Device`s.
+
+## Sending Push Notifications
+
+`Notification`s are structured as follows:
+
+- **alert**: Displayed message on the device which includes title, subtitle, and body.
+- **payload**: Service specific notification settings like sound, badges, images, etc.
+- **data**: Background data.
+
+`Notification`s don't have to include an alert and can just send a silent push notification with background data.
+
+```swift
+public struct Notification {
+    public var alert: Alert?
+    public var payload: Payload?
+    public var data: Encodable?
+}
+```
+
+Example of a `Component` that sends a push notification to an user:
+
+```swift
+struct SendNotification: Component {
+    @Request(\.user) var user: User
+
+    @Environment(\.notificationCenter) var notificationCenter: NotificationCener
+
+    func handle() -> EventLoopFuture<Void> {
+        notificationCenter.send(alert: .init(title: "Hello There 👋"), to: user)
+    }
+}
+```
+
+The `NotificationCenter` offers the following convenience methods to send push notifications:
+
+- Send to all registered devices
+- Send to a subscription
+- Send to all devices of a user
+- Send to multiple users
+- Send to a specific device
+- Send to a specific token
+
+## Unsolicited Push Notifications
+
+Events in Apodini can trigger the sending of push notifications. `Component`s can subscribe to `ObservableObject`s or **[CronJobs](./CronJob.md)** that are exposed to the environment. The `handle` method of a `Component` will be executed every time `@Published` properties of the `ObservableObject` or `CronJob` change. Unsolicited `Component`s are defined in Apodini by conforming to the `EventComponent` protocol. This specifies that `EventComponent`s are only executed by events which the `EventComponent` subscribes to, whereas incoming requests of clients will not be handled. Therefore, property wrappers that are request based, e.g. `@Request` or `@Parameter`, cannot be used in an `EventComponent` and will throw an error. This also means that Apodini exporters will ignore this `Component`.
+
+The following example shows a _WeatherService_ as a `CronJob` that fetches the weather every day at 10 am and publishes it to its subscribers using the `@Published` property wrapper. The _AlertComponent_ listens to changes of the _WeatherService_ and sends out push notifications based on the current weather. Every `Device` with the **subscription** _weatherSubscription_ will receive this push notification.
+
+```swift
+struct WeatherService: CronJob {
+    @Published var weather = "sunny"
+
+    func run() {
+        // Updates weather
+    }
+}
+
+struct AlertComponent: EventComponent {
```

Comment:
While I see the need and the requirement to enable unsolicited request I think that the idea of marking them as `Component`s or `Handler`s might be a bit misleading. Would it be possible to include all that functionality directly into the `CronJob`? So a `CronJob` would have a `run` method and would be able to store state using `@State` or even observe other objects using `@ObservedObject` or access the environment using `@Environment`. By accessing the environment a `CronJob` could also send out `Notification`s using the `NotificationCener` and could even update other global `@ObservedObject` that might trigger different other components. 

Alternatively they might also be able to directly trigger service or bidirectional streams by behaving like `ObservedObjects`. These two modes under one hood could eliminate the need for `EventComponent `s and would make the act of adding a `Component` or `Handler` in a `content` property that is ignored by all exporters nonexistent.

This functionality could then be described in the CronJobs documentation.

Reply:
Yes, `CronJob`s can also use `@Environment` and I think we can extend this also to other property wrappers such as `@State`. The only concern I have is using `@ObservedObject` to listen to another object inside of a `CronJob`. This might be confusing because the `run` method would be executed everytime the `CronJob` is scheduled and is also triggered if one `@ObservedObject` changed. That is why I included the `EventComponent`. However, I like the idea that `CronJob`s can update properties of global `@ObservedObject`s.

I agree with you that other `Component`s should use `@ObservedObject` or `@Environment` to subscribe to changes of `CronJob`s, which can be used for server-side streams or bidirectional communication to a client.

Reply:
@awoc I would argue that it would even be fine if an `@ObservedObject` can trigger an execution of the `run` function even though I only see very specific use cases for this at the moment. E.g. an Edge Chase of a `CronJob` that would fire once or never and would just be used to observe other `@ObservedObject` without the need to register and clutter the `Component`s DSL as this feature should not be triggered by Requests. Or a `CronJob` that would fire every 24 hours to e.g send a summary of the day or if the user manually requests such a report. 

Maybe It would even make sense to generalize the feature to `Job`s instead of `CronJob`s as cron would be very specific feature to configure them but I could also image some other ways to configure them except for cron Strings.

I think we would need a mechanism to check what has fired. This would not only apply to `Job`s but also to `Handler`s. A sort of way to access the projected value of a property wrapper to check if it has triggered the `handle` method in a `Handler` or the `run` method in a `Job`. E.g.:
```swift
struct ExampleJob: Job {
    @Environment(\.logger) var logger: Logger
    @ObservedObject var test: SomeObject = SomeObject()

    func run() {
        if $test.changed {
            logger.log("The `@ObservedObject` changed.")
        } else {
            logger.log("The Job was triggered.")
        }
    }
}
```
That would enable the developer to differentiate what has triggered the `Handler` or `Job` in the edge cases where this information is important.

Reply:
I like the idea to check if an `@ObservedObject` has triggered. This would be especially useful when using multiple `@ObservedObject`s in a `Component` in order to differentiate them.

Reply:
Maybe we can incorporate this idea in the PR so we can further refine and specify the behavior of `@ObservedObject`  and `@Environment` with this `.changed` behavior? I would see that `Job`s could benefit from this behavior and I agree that `Handler`s could also greatly benefit from this feature.

Reply:
If `Job`s should also have the ability to change properties of `ObservedObjects`, I think it would be good to differentiate between the property wrappers `@Environment` and `@ObservedObject` to have a different meaning. I could envision that `Components` or `Handlers` can change the state of `ObservedObject`s with `@Environment` but don't listen to events that are emitted. On the other hand, the property wrapper `@ObservedObject` is used when `Component`s listen to events which will trigger the `handle` method everytime one property annotated with `@Published` changes. The same would apply to `Job`s. This would only work for globally defined `ObservedObject`s or when declared to child `Component`s with the modifier `.environment()`.

Let's consider the following example: The component `RegisterVisitor` counts every person that visits this specific path and saves this in `VisitorObject`. Everytime `VisitorObject` changes its property `count` the `Summary` job is executed and sends out a push notification. Furthermore, the `Summary` job also sends a summary of all visitors every Friday at 9 am. Global `ObservedObject`s are declared in the `configuration` property with the `Environment` configuration. This would define a key path which can either be used in every `Component` or `Job` using the `@ObservedObject(\.keypath)` or `@Environment(\.keypath)`.


```swift
struct VisitorObject: ObservedObject {
    @Published var count = 0
}

struct Summary: Job {
    @ObservedObject(\.visitorObject) var visitorObject: VisitorObject

    @Environment(\.notificationCenter) var notificationCenter: NotificationCenter

    func run() {
        if $visitorObject.changed {
            notificationCenter.send(alert: Alert(title: "We have a new customer"), to: "visitorTopic")
        } else {
            notificationCenter.send(alert: Alert(title: "This week we had a total of \(visitorObject.count) visitors", to: "visitorTopic"))
        }
    }
}

struct RegisterVisitor: Component {
    @Environment(\.visitorObject) var visitorObject: VisitorObject

    func handle() -> String {
        visitorObject.count += 1
        return "Welcome to our site!"
    }
}

var content: some Component {
    RegisterVisitor()
}

var configuration: Configuration {
    Schedule(Summary.self, on: "0 9 * * 5")
    Environment(\.visitorObject, VisitorObject())
}
```

Reply:
I like this differentiation and I think this sounds reasonable!
The option of have an `@ObservedObject` with a KeyPath to observe an object in the environment and without a key path to observe an object that is created locally or passed in using the initializer creates a nice API. And it makes it very clear if an object should be observed (`@ObservedObject`) and just used (`@Environment`).

The example sounds good and I would suggest to extend the example to include the base Apodini `WebService` and then add it as part of the PR in the documentation folder. 
There are a lot of different ideas in this review thread and it would make sense to summarize everything and as part of documentation files, update the PR and send the PR out to the whole team to give everyone the change to review the changes and get familiar with the proposed ideas 👍